### DISCUSSION TITLE
How to remove antialias edge when converting to a binary bitmap?

### DISCUSSION THREAD

**Discussion by pzelchenko** [ImageMagick/ImageMagick#4572] [Help] [answered]
![image](https://user-images.githubusercontent.com/7409590/145495090-9aaf0442-d3be-4789-bdec-b771e0a71e6a.png)
ImageMagick is an incredibly versatile product. It's the only thing I can use for this kind of work. I'm doing a cognitive experiment that requires flashing 480 black graphics at high speed against a neutral-gray background. The image shown here is enlarged detail of output from ImageMagick; the full images are 932 x 612 pixels, and there are actually other clusters similar to this one elsewhere in the image.

I have to generate these images in a source app and later display them in the experimental app. The source app can only generate 24-bit PNG's with gray or white background (see [white](https://pete.zelchenko.com/2021/e001-white.png) and [gray](https://pete.zelchenko.com/2021/e001-gray.png) example source images). Despite the size of each of these files hovering around 10KB, the total file payload of 480 files is too large and takes way too long to preload. Although I can change the background color in the source application, I don't have control over bit depth there, so we are stuck generating screen captures and then trying to downsample in ImageMagick. I can set a default background color in the experimental app and put up 1-bit images. Using ImageMagick, some things I've tried:

`magick e001-gray.png -transparent "rgb(192,192,192)" -alpha set -depth 1 out\gray-out.bmp` (or `.png`)
`magick e001-white.png -transparent white -alpha set -depth 1 out\white-out.bmp` (or `.png`)

I have tried the above commands with original images that sit on both a white background and the gray background. I start with files of around 10KB. Most trials give me files of around 4K. The only thing that gets a file size of <2K, oddly, is not specifying -transparent rgb(255,255,255), but anything else works, whether I start with a gray- or white-background image. The problem is that all of these kinds of attempts exhibit the artifact shown in the image above. Increasing the bit depth can make it disappear, but then the file size is too large.

I've also tried saving as GIF without -transparent but with -edge, and PNG with -transparent rgb(192,192,192) but without -edge. Both make files <2 KB. But the -edge 1 makes a blank file, and the -transparent still exhibits the artifact.

Is there any way to get tiny files with a transparent background, without that artifact?

**Comment by GeeMack** [comment]
Using ImageMagick v7, using your example image, and running this command...

    magick sample.png -background none -threshold 50% -transparent white output.png

I get this result (transparent background appears white)...

![output](https://user-images.githubusercontent.com/65475279/145500203-e3a65051-0c1f-4426-b860-b55941b4bdbc.png)

The command uses `-threshold 50%` to make everything brighter than 50% white, and everything darker than 50% black. Then it uses `-transparent white` to turn all the white to transparent.

With my IMv7 that output PNG file is only 1,198 bytes.

**Comment by snibgo** [comment]
As you realise, your input images are antialised. This means they have intermediate shades of gray between black and white (or black and gray).

Your commands make only one colour transparent: the white or the gray. So you leave the other shades of gray alone. Is that what you mean by "the artifact shown in the image above"?

If so, then an obvious solution is remove the intermediate shades, making the image black and white only, before making white transparent. For example (this works on both your linked images):
```
magick e001-gray.png -alpha off -auto-level -transparent white -background Black -alpha background -alpha set -depth 1  out.png
```

> Most trials give me files of around 4K. The only thing that gets a file size of <2K, oddly, is not specifying -transparent rgb(255,255,255), ...

That isn't odd. Without transparency, there is one channel. With transparency, there are two channels. When you double the channels, you can expect double the filesize.

**Comment by pzelchenko** [comment]
Thanks, guys. The first solution, by @GeeMack, is giving me only a totally transparent file with no visible content (the thumbnail looks black). This happens whether I use the white or the gray original. The second solution, by @snibgo, is transparent but still showing that same artifact of white around the edge. I'm using ImageMagick 7.1.0-17.

**Comment by snibgo** [comment, marked as answer]
Yes, sorry, I forgot the "-threshold":
```
magick e001-gray.png -alpha off -auto-level -threshold 50% -transparent white -background Black -alpha background -alpha set -depth 1  x.png
```

**Reply by pzelchenko** [reply]
Well, the above worked great. This file is 3.77 KB:
![image](https://user-images.githubusercontent.com/7409590/145556128-dca5e87a-890b-4328-a35f-b57cc3e70a45.png)
There seems to be a lot going on in that command, @snibgo. I see three -alpha flags which suggests that this is more like a script than a set of static arguments, something I didn't know about ImageMagick but makes perfect sense. Let me see if I understand correctly:

 - `-alpha off` first turns off existing alpha channel;
 - `-auto-level` then stretches the histogram to 0-255;
 - `-threshold 50%` forces anything over the middle of the histogram to black;
 - `-transparent white` treats all "white" pixels as transparent (actually, any pixel below the 50% threshold, which explains why this works for white _and_ gray?);
 - `-background Black` is literally self-explanatory, but I don't quite see why you use black in particular -- is it so that:
 - `-alpha background` will then minimize the file size on compression (black = 0x000000)?;
 - `-alpha set` activates the alpha channel, and
 - `-depth 1` of course makes it a 1-bit image.

The resulting file should contain a bitmap and an alpha channel.

I've played around with various formulas and keep coming up with a file of about 3.8 KB. But some files are much smaller. What if all of the information could be transfered to the alpha channel, and the main bitmap were made solid black? Wouldn't that give the ideal compression?

**Comment by snibgo** [comment]
IM commands can be massively complex, reading and writing thousands of files, with thousands of operations. IM has a script language (well, two or more languages, depending on how we count them). And IM can be run from a shell script (bash, Windows BAT, or whatever).

> What if all of the information could be transfered to the alpha channel, and the main bitmap were made solid black? Wouldn't that give the ideal compression?

My command does that.

Inserting "-strip" after reading your input removes about 200 bytes. "pngcrush -brute -v x,png x1.png" removes another 700 bytes, so x1.png is 3099 bytes. If we don't use transparency so the image is just black and white, the size is 1772 bytes, reduced to 1672 by pngcrush.

See https://imagemagick.org/script/command-line-options.php for an explanation of each option.

"-strip": remove metdata that doesn't affect the image

"-auto-level": stretch the pixel values to be 0 to QuantumRange. For IM Q8, QuantumRange is 255. For your grayscale image, this makes the lightest pixels white and the darkest pixels black.

"-threshold 50%": for your image, pixels lighter than 50% are made white. Pixels darker than or equal to 50% are made black.

"-transparent white": make white pixels transparent. So now the image has opaque black and transparent white only.

"-background Black -alpha background": make all fully-transparent pixels into transparent black. So now all the pixels are black, with alpha either transparent or opaque.

"-alpha set": I've forget why I did this. Not needed on your images. It might be needed on some images.

"-depth 1": when the image is saved, IM should use 1 bit per pixel.

**Reply by pzelchenko** [reply]
It seems that if I want transparency, the file will be approximately twice the size (total around 3K), due to the alpha channel. And I now somewhat understand your syntax for setting the main image to a full field of black and the data is now in the alpha channel.

What I don't understand is, if the main image has essentially reduced to a run of zeroes and is a 1 bit image, then 932 x 612 x 1 bit = 570384 bits / 8 ~= 69.62 KB uncompressed. But then shouldn't the main image then compress down to almost nothing? I created an all-black monochrome PNG in mspaint of those dimensions and it's 217 bytes, and pngcrush only removed 3 additional bytes. If a pure two-color image (or, presumably, an alpha channel) is 1772 bytes (as you say above), then shouldn't the all-black field plus the alpha channel be not much more than 2000 bytes? What else is there to nearly double the file size?

My suspicion is that the -depth 1 is not reducing it to 1 bit. A 1-bit image with an alpha channel doesn't make sense, so perhaps the whole PNG is actually 8 bits, and although the main image does reduce down, the 8-bit alpha channel is going to be around 2.5K?

I really appreciate @snibgo and also @GeeMack for your time on this. I know this is not free. Pizza's in the oven if you can make it to Wuxi, China.

**Comment by snibgo** [comment]
I see what you mean about compression, but I don't know enough about PNG internals to comment.

If filesize is critical for you, I suggest you experiment with other file formats such as gif, webp, jp2 and tiff.

You might also take a wider view: the application that displays the images to your viewers logically needs only one channel, one bit per pixel. Perhaps it can be modified to take input files that have only one channel, one bit per pixel.

**Reply by pzelchenko** [reply]
It appears that the PNG is still 8 bits.

>magick identify snib01.png
snib01.png PNG 932x612 932x612+0+0 8-bit Grayscale Gray 3811B 0.000u 0:00.016

http://www.libpng.org/pub/png/book/chapter08.html

"8.5.4. Grayscale with Transparency
PNG supports two kinds of transparency with grayscale and RGB images. The first is a palette-style ``cheap transparency,'' in which a single color or gray value is marked as being fully transparent. I noted earlier that the structure of tRNS depends on the image type; for grayscale images of any pixel depth, the chunk contains a 2-byte, unscaled gray value--that is, the maximum allowed value is still 2bit depth-1, even though it is stored as a 16-bit integer. This approach is very similar to GIF-style transparency in palette images and incurs only 14 bytes overhead in file size. There is no corresponding TIFF image type, and standard JPEG does not support any transparency."

I guess one strategy would be to unlock that "cheap" 14-byte transparency in the file after removing the antialiasing.

Anyhow, I think we can say that your first response answers the basic question and the rest is more arcane. Thanks again. If you have any additional insights, I'd welcome them.

**Reply by pzelchenko** [reply]
It looks as if my colleague is running the experiment on PsyToolkit (in England) from China. PsyToolkit by default has the user upload files to its own server. So the issue appears to be total payload (300 similar files = ) into China, which as individual files slows to a crawl. PsyToolkit supports a `house http://mysite.com/image.png` command (although unfortunately it doesn't support a simple `file:///c$/` syntax). We're going to try setting up a localhost server.

**Comment by snibgo** [comment]
If network speed is the bottleneck, another possibility is to ensure the files have one channel, one bit before upload (pixels are black and white only), and the receiver then converts the files as required, eg making white pixels transparent.

> ... flashing 480 black graphics at high speed ...

Another possibility is to create the images as an animation or video.

**Reply by pzelchenko** [reply]
Thanks for the ideas. Actually, I'm checking with the developer of PsyToolkit now to see why your one-bit option is not supported. That way, it would go down to about 2 KB per file.

The video is not a bad idea, except I think we'd simply be encountering the same problem, only on a frame basis. The main problem with that is that this is a cognitive psychology experiment, and that requires highly synchronized timing of blanking the screen, presentation of the next screen of stimulus, wait for input, immediate reloop at high speed.

Best option right now is to get the files small (that's always a good policy in any case), and get it on a local server. Thanks again.
